### PR TITLE
Add an HeartbeatStandalone strategy

### PR DESCRIPTION
**Pull request by gaetanmaisse**
## Issue
https://github.com/gravitee-io/issues/issues/8482

## Description

Refactor HeartbeatService to split implementation into two strategies.
Current implementation is moved to `HazelcastHeartbeatStrategy` 
Create a dedicated standalone implementation: `StandaloneHeartbeatStrategy`

It is responsible of creating or updating the heartbeat event at fixed interval.

Once the event has been created, it has been decided to call the `createOrUpdate` method with a lite version of the event, containing only the requested field for the update (id`,` `updatedAt`, `properties.last_heartbeat_at`)

### Why ?

Heartbeat event is a big object (it contains the list of plugins, and a big list of properties).

Sending only the required fields for update allow to:
- save bandwidth between gateway client and gateway server
- save bandwidth between gateway server and dbms
- by default, event are sent every 5 seconds, for x gateways running, it can be a lot of event to handle at nearly the same time for the dbms, reducing the size reduce the pain.
- query are now optimized for update

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-mpxsovtykb.chromatic.com)

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/fix-heartbeat-service-3-18-x/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatStrategy.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.gateway.services.heartbeat;
+
+import io.gravitee.repository.management.model.Event;
+import java.util.function.Supplier;
```
Comment (Okhelifi) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatStrategy.java]: not used 🙈

**Review by Okhelifi — APPROVED** [review]

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatStrategy.java (hunk 1 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatStrategy.java]: force pushed

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatService.java [hunk 2]
```diff
@@ -253,6 +134,41 @@ private Event prepareEvent() throws TechnicalException {
...
+            environments = environmentRepository.findByOrganizationsAndHrids(organizationsIds, environmentsHrids);
+        }
+
+        Set<String> environmentsIds = environments.stream().map(Environment::getId).collect(Collectors.toSet());
+
+        // The first time APIM starts, if the Gateway is launched before the environments collection is created by the Rest API, then environmentsIds will be empty.
+        // We must put at least "DEFAULT" environment.
+        if (environmentsIds.isEmpty()) {
+            environmentsIds.add("DEFAULT");
+            environmentsHrids = environmentsHrids.isEmpty() ? Collections.singleton("DEFAULT") : environmentsHrids;
```
Comment (marcambier) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatService.java:161]: Could we better use the `GraviteeContext.getDefaultEnvironment()` constant here ?

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java [hunk 3]
```diff
@@ -280,6 +280,47 @@ public List<Event> search(EventCriteria filter) {
         return queryEvents(builder.toString(), args);
     }
 
+    @Override
+    public Event createOrUpdateHeartbeat(Event event) throws TechnicalException {
+        if (event == null || event.getId() == null) {
+            throw new IllegalStateException("Event to create or update must have an id");
+        }
+
+        if (existsById(event.getId())) {
```
Comment (marcambier) [inline gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java]: Maybe we could get rid of `existsById`,
Just try to update, and create if nothing was updated (updatedEventCount < 1).
That would save a query, and so, improve performance. But maybe it's a little dirty.
WDYT ?

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java [hunk 4]
```diff
@@ -280,6 +280,47 @@ public List<Event> search(EventCriteria filter) {
...
+        return create(event);
+    }
+
+    private int updateHeartbeatEventDate(Event event) {
+        return jdbcTemplate.update("update " + this.tableName + " set updated_at = ? where id = ?", event.getUpdatedAt(), event.getId());
+    }
+
+    private int updateHeartbeatEventProperties(Event event) {
+        return jdbcTemplate.update(
+            "update " + EVENT_PROPERTIES + " set property_value = ? where event_id = ? and property_key = 'last_heartbeat_at'",
```
Comment (marcambier) [inline gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java]: As a performance improvment, is it relevant to add an index on event_id + property_key, according to the new query here ?
I see the index `idx_${gravitee_prefix}eventproperties_propertykey` for property_key, but it doesn't include id.

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/test/java/io/gravitee/gateway/services/heartbeat/impl/standalone/StandaloneHeartbeatStrategyTest.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    public static final String ENVIRONMENTS_PROPERTIES = "environments";
+
+    @Mock
+    private EventRepository eventRepository;
+
+    @Mock
+    private HeartbeatDependencies heartbeatDependencies;
+
+    @Captor
+    ArgumentCaptor<Event> eventCaptor = ArgumentCaptor.forClass(Event.class);
```
Comment (marcambier) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/test/java/io/gravitee/gateway/services/heartbeat/impl/standalone/StandaloneHeartbeatStrategyTest.java]: I think the @Captor makes the instantiation useless.
```suggestion
    ArgumentCaptor<Event> eventCaptor;
```

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/spring/configuration/HeartbeatDependencies.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.gravitee.gateway.env.GatewayConfiguration;
+import io.gravitee.node.api.Node;
+import io.gravitee.plugin.core.api.PluginRegistry;
+import io.gravitee.repository.management.api.EnvironmentRepository;
+import io.gravitee.repository.management.api.EventRepository;
+import io.gravitee.repository.management.api.OrganizationRepository;
+import java.util.concurrent.TimeUnit;
+
+public class HeartbeatDependencies {
```
Comment (marcambier) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/spring/configuration/HeartbeatDependencies.java]: I'm a little confused about the `dependencies` naming.
It looks like a configuration class to me.
Naming suggestion : `HeartbeatStrategyConfiguration`

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/impl/standalone/StandaloneHeartbeatStrategy.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            );
+        }
+    }
+
+    /**
+     * Provides a lite version of the Heartbeat Event to only update date related fields
+     * @param heartbeatEvent the original heartbeat event
+     * @return the lite database
+     */
+    private Event getLiteEvent(Event heartbeatEvent) {
```
Comment (marcambier) [inline gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/impl/standalone/StandaloneHeartbeatStrategy.java]: The `get` naming makes me think it retreives an event.
But at all, this just build an event object.
Naming suggestion : `buildLiteEvent`

**Review by marcambier — COMMENTED** [review]
Nice job :+1: 
Some minor remarks here after reading the code, before starting testing.
Nothing important, feel free to ignore my comments if they are not relevant

File: gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatService.java (hunk 2 above)
Comment (ytvnr) [reply gravitee-apim-gateway/gravitee-apim-gateway-services/gravitee-apim-gateway-services-heartbeat/src/main/java/io/gravitee/gateway/services/heartbeat/HeartbeatService.java:161]: I think it only exists in rest-api and not in gateway, wdyt ?

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java (hunk 3 above)
Comment (ytvnr) [reply gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java]: yes it's a good point, I will propose a commit

File: gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java (hunk 4 above)
Comment (ytvnr) [reply gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcEventRepository.java]: discussed

**Review by marcambier — APPROVED** [review]

**Pull request merged**